### ISSUE TITLE
Misleading documentation: port number can be outside of 1-9 range

### ISSUE THREAD

**Issue by balsoft** [TheWeirdDev/Bluetooth_Headset_Battery_Level#33]
As stated in the readme:

> Try port numbers 1 to 9 to find the one that works for your device. (wait a few seconds between each try)

However, my WH-1000XM3 works when I specify port number 10 (and doesn't work for any port in 1-9 range). CC @chaoscreater BTW, try disconnecting, reconnecting and then running `./bluetooth_battery <MAC>.10` . I suspect this may also be true for other headsets.

**Comment by bhepple** [comment]
Ha! Me too! I have an Audeara A-01 and port 10 is the ace.

**Comment by chaoscreater** [comment]
Just tried it on my PC at home. Still not working for me. The MAC address is definitely correct.

![image](https://user-images.githubusercontent.com/18227319/89376071-34647180-d743-11ea-9d38-9bf773ddb578.png)

**Comment by TheWeirdDev** [comment]
@balsoft Thanks for letting me know.
 I couldn't find useful information about Bluetooth port numbers and decided to go with 1-9 because i hadn't seen any larger numbers being used in other apps. I'll update the readme.

**Comment by TheWeirdDev** [comment]
> Just tried it on my PC at home. Still not working for me. The MAC address is definitely correct. Have tried disconnecting and reconnecting.
> 
> ![image](https://user-images.githubusercontent.com/18227319/89376071-34647180-d743-11ea-9d38-9bf773ddb578.png)

Try running the script while your device is completely disconnected. And perhaps try larger port numbers, maybe 10 to 20 to see if those will work.

**Comment by balsoft** [comment]
@chaoscreater 

> Just tried it on my PC at home. Still not working for me. The MAC address is definitely correct. Have tried disconnecting and reconnecting.

The error is not like anything I've seen before. I suspect that there is another issue you're facing. To be clear, I'm running Linux, so maybe there's some differences here. Maybe it's because of the whole `LE_...` business? Can you try connecting to the headset so that it shows up as just `WH-1000XM3`?

**Comment by chaoscreater** [comment]
LE stands for low energy. I'm not really sure how to put it into standard mode. When I paired my XM3, it just paired as WH-1000XM3 and sometimes it will switch to LE. I think this is normal, I've seen this on my Bose QC35 II as well and on different machines. But yeah I'll try play around with the settings a bit.

**Comment by balsoft** [comment]
> OK, it's working now. I haven't done anything different. I just keep trying to disconnect my WH-1000XM3 from Windows settings and running the script with a different port from 10~20. It somehow worked on 10 even though it wasn't working before.

The magic of bluetooth :shrug:

This confirms that 10 is correct port, but there's clearly something... Interesting going on there.

Just FYI, it works consistently for me.

**Comment by bhepple** [comment]
In fact, multiple ports appear to work for me on my Audeara A-01 - .10 as reported above but also .12!! Lucky me.

I discovered the port .10 while fiddling with another script at https://github.com/ThePBone/GalaxyBuds-BatteryLevel (my version is below)

It uses hard-coded **uuids** presumably specific to the headset. I discovered mine using _blueman-manager_ (on linux). By right clicking on the headset entry and then 'Info' it showed the uuids that the headset knows about:

```
00001101-0000-1000-8000-00805f9b34fb Serial Port
0000110b-0000-1000-8000-00805f9b34fb Audio Sink
0000110c-0000-1000-8000-00805f9b34fb Remote Control Target
0000110d-0000-1000-8000-00805f9b34fb Advanced Audio
0000110e-0000-1000-8000-00805f9b34fb Remote Control
0000111e-0000-1000-8000-00805f9b34fb Handsfree
00001200-0000-1000-8000-00805f9b34fb PnP Information
```

I then tried these until I found one that responded in my modified buds_battery.py script - the magic one was the uuid for 'Handsfree'. It showed me a data block:
```

$ ./buds_battery.py -v F0:23:B9:50:CA:54 
Checking device model...
Audeara A-01
Searching for the RFCOMM interface...
uuid =  0000111e-0000-1000-8000-00805f9b34fb
service_matches =  [{'service-classes': ['111E', '1203'], 'profiles': [('111E', 263)], 'name': 'Hands-Free unit', 'description': None, 'provider': None, 'service-id': None, 'protocol': 'RFCOMM', 'port': 12, 'host': 'F0:23:B9:50:CA:54'}]
RFCOMM interface found. Establishing connection...
Connected. Waiting for incoming data...
b'AT+BRSF=959\r'
```

... initially it gave me 'port':10 - now it's giving 'port':12

Maybe something here would be useful in probing a general-case headset and discovering the port?

Here's the modified buds_battery.py:

```
#!/usr/bin/env python3

"""
A python script to get battery level from Samsung Galaxy Buds(+) devices
"""

# License: MIT
# Author: @ThePBone
# 06/30/2020

import bluetooth
import sys
import argparse
import datetime

msg_debounce = ""


def print_result(string, timestamp):
    global msg_debounce
    if msg_debounce != string:
        if timestamp:
            print(string + "," + datetime.datetime.now().strftime("%FT%T.%f")[:-4] + "Z")
        else:
            print(string)
        msg_debounce = string


def parse_message(data, isplus, timestamp):
    if data[0] != (0xFD if isplus else 0xFE):
        print("Invalid SOM")
        exit(2)
    if data[3] == 97:
        if isplus:
            string = "{},{},{}".format(data[6], data[7], data[11])
        else:
            string = "{},{}".format(data[6], data[7])
    elif data[3] == 96:
        if isplus:
            string = "{},{},{}".format(data[5], data[6], data[10])
        else:
            string = "{},{}".format(data[5], data[6])
    else:
        return False
    print_result(string, timestamp)
    return True


def parse_message_wear_status(data, isplus, timestamp):
    global msg_debounce
    if data[0] != (0xFD if isplus else 0xFE):
        print("Invalid SOM")
        exit(2)
    if data[3] == 97:
        state = data[10]
    elif data[3] == 96:
        state = data[9]
    else:
        return False

    if isplus:
        left = id_to_placement((state & 240) >> 4)
        right = id_to_placement(state & 15)
        string = "{},{}".format(left, right)
    else:
        if state == 0:
            string = "None"
        elif state == 1:
            string = "Right"
        elif state == 16:
            string = "Left"
        elif state == 17:
            string = "Both"
        else:
            string = "Unknown"

    print_result(string, timestamp)
    return True


def id_to_placement(id):
    if id == 0:
        return "Disconnected"
    elif id == 1:
        return "Wearing"
    elif id == 2:
        return "Idle"
    elif id == 3:
        return "InCase"
    elif id == 4:
        return "InClosedCase"


def main():
    parser = argparse.ArgumentParser(description='Read battery values of the Samsung Galaxy Buds or Buds+ '
                                                 '[Left, Right, Case (Buds+)]')
    parser.add_argument('mac', metavar='mac-address', type=str, nargs=1,
                        help='MAC-Address of your Buds')
    parser.add_argument('-m', '--monitor', action='store_true', help="Notify on change")
    parser.add_argument('-t', '--monitor-timestamp', action='store_true', help="Notify on change and print timestamps")
    parser.add_argument('-w', '--wearing-status', action='store_true', help="Print wearing status instead")
    parser.add_argument('-v', '--verbose', action='store_true', help="Print debug information")
    args = parser.parse_args()

    verbose = args.verbose

    if verbose:
        print("Checking device model...")
    isplus = "Buds+" in str(bluetooth.lookup_name(args.mac[0]))
    if verbose:
        print(str(bluetooth.lookup_name(args.mac[0])))

    if verbose:
        print("Searching for the RFCOMM interface...")

    #uuid = "00001101-0000-1000-8000-00805f9b34fb"
    #uuid = "0000110b-0000-1000-8000-00805f9b34fb"
    #uuid = "0000110c-0000-1000-8000-00805f9b34fb"
    #uuid = "0000110d-0000-1000-8000-00805f9b34fb"
    #uuid = "0000110e-0000-1000-8000-00805f9b34fb"
    uuid = "0000111e-0000-1000-8000-00805f9b34fb"
    #uuid = "00001200-0000-1000-8000-00805f9b34fb"

    #uuid = ("00001101-0000-1000-8000-00805F9B34FB" if isplus else "00001102-0000-1000-8000-00805f9b34fd")
    service_matches = bluetooth.find_service(uuid=uuid, address=str(args.mac[0]))

    if len(service_matches) == 0:
        print("Couldn't find the proprietary RFCOMM service")
        sys.exit(1)

    if verbose:
        print("uuid = ", uuid)
        print("service_matches = ", service_matches)

    port = service_matches[0]["port"]
    host = service_matches[0]["host"]

    if verbose:
        print("RFCOMM interface found. Establishing connection...")
    sock = bluetooth.BluetoothSocket(bluetooth.RFCOMM)
    sock.connect((host, port))

    if verbose:
        print("Connected. Waiting for incoming data...")

    try:
        while True:
            data = sock.recv(1024)
            if len(data) == 0:
                break
            print(data)
            sock.send(b"+BRSF: 1024")
            sock.send(b"OK")
            data = sock.recv(1024)

            if args.wearing_status:
                success = parse_message_wear_status(data, isplus, args.monitor_timestamp)
            else:
                success = parse_message(data, isplus, args.monitor_timestamp)

            if success and not args.monitor and not args.monitor_timestamp:
                exit(0)

    except IOError:
        pass


if __name__ == "__main__":
    main()
```

**Comment by TheWeirdDev** [comment]
@bhepple That's great!
I have added a new experimental branch that works with this code. Please try it and let me know if it works.

**Comment by balsoft** [comment]
@TheWeirdDev the branch works for me:

```
$ python3 ./bluetooth_battery.py CC:98:8B:C0:FC:D2
Battery level for CC:98:8B:C0:FC:D2 is 80%
```

Thanks a lot!

**Comment by TheWeirdDev** [comment]
Thanks to both of you for your help ❤️. It's now merged into the master branch.

**Comment by bhepple** [comment]
Yep - the latest version works for me too! Well done.

**Comment by chaoscreater** [comment]
I've tried the latest version. Still got the same issue as I mentioned, where it will detect the battery (only if I choose port 10) and then it will disconnect the bluetooth. Essentially, every port I try will successfully make the bluetooth connection, but only port 10 will show the battery level and it will just disconnect straight away.

I've tried uninstalling bluetooth_battery and reinstalling and make sure I'm on the latest version. Same issue.
